platform/pulp/plugin/changeset/main.py

```diff
@@ -0,0 +1,219 @@
+            with ProgressBar(message=_('Add Content'), total=len(self.additions)) as bar:
+                for plan in batch():
+                    artifact = plan.download.attachment
+                    artifact.model.downloaded = True
+                    content = artifact.content
+                    if not plan.succeeded:
+                        task = Task()
+                        task.append_non_fatal_error(plan.error)
+                        bar.increment()
+                        report = ChangeReport(ChangeReport.ADDED, content.model)
+                        report.errors.append(plan.error)
+                        yield report
+                        continue
+                    content.settle()
+                    if not content.settled:
+                        continue
+                    with transaction.atomic():
+                        self._add_content(content)
+                        bar.increment()
+                    report = ChangeReport(ChangeReport.ADDED, content.model)
+                    yield report
+
+    def _apply_removals(self):
+        """
+        Apply removals.
+        Content listed in `removals` is removed from the repository.
+
+        Yields:
+            ChangeReport: A report for each removed content.
+        """
+        log.info(
+            _('Apply removals: repository=%(r)s.'),
+            {
+                'r': self.repository.name
+            })
+        with ProgressBar(message=_('Remove Content'), total=len(self.removals)) as bar:
+            for batch in BatchIterator(self.removals, 100):
+                with transaction.atomic():
+                    for content in batch:
+                        self._remove_content(content)
+                        bar.done += len(batch)
+                        bar.save()
+                        report = ChangeReport(ChangeReport.REMOVED, content)
+                        yield report
+
+    def apply(self):
+        """
+        Apply the changes to the repository based on download policy.
+
+        Yields:
+            ChangeReport: For each change applied.
+        """
+        log.info(
+            _('Apply ChangeSet: repository=%(r)s.'),
+            {
+                'r': self.repository.name
+            })
+        for report in self._apply_additions():
+            yield report
+        for report in self._apply_removals():
+            yield report
+
+
+class SizedIterable(Sized, Iterable):
+    """
+    A sized iterable.
+
+    Attributes:
+        iterable (Iterable): An iterable.
+        length (int): The number of items expected to be yielded by the iterable.
+    """
+
+    def __init__(self, iterable, length):
+        """
+        Args:
+            iterable (Iterable): An iterable.
+            length (int): The number of items expected to be yielded by the iterable.
+        """
+        self.iterable = iterable
+        self.length = length
```

Comment:
+1 to at least make attributes private/protected so it won't be easy to overwrite them.